composer/optim/scheduler.py

```diff
@@ -1,226 +1,265 @@
 
-_interval_doc = 'frequency of step() calls, either "batch" or "epoch". Default: "epoch"'
-
-# Allow (batch, batches) or (epoch, epochs). Also accept "step" ~ "batch"
-INTERVAL_MAP = {
-    'batch': 'batch',
-    'batches': 'batch',
-    'epoch': 'epoch',
-    'epochs': 'epoch',
-    'step': 'batch',
-    'steps': 'batch'
-}
-
-
-def _convert_time_fields(interval: str,
-                         kwargs: Dict[str, Any],
-                         max_training_duration: Optional[Union[str, Time[int]]] = None,
-                         steps_per_epoch: Optional[int] = None,
-                         samples_per_epoch: Optional[int] = None,
-                         dataset_num_tokens: Optional[int] = None) -> None:
-    """Converts all fields in ``kwargs`` that were provided as timestrings (e.g. "32ep") into integers, representing
-    either epochs or batches, depending on the ``interval``.
-
-    Modifies ``kwargs`` in place.
-    """
-    interval_unit = TimeUnit(INTERVAL_MAP[interval])
-
-    for field_name, field_value in kwargs.items():
-
-        if field_name not in ('interval', 'warmup_method'):
-            if isinstance(field_value, list) and all(isinstance(x, str) for x in field_value):
-                kwargs[field_name] = [
-                    convert_time(t,
-                                 unit=interval_unit,
-                                 steps_per_epoch=steps_per_epoch,
-                                 max_training_duration=max_training_duration,
-                                 samples_per_epoch=samples_per_epoch,
-                                 dataset_num_tokens=dataset_num_tokens).value for t in field_value
-                ]
-                continue
-            if isinstance(field_value, str):
-                kwargs[field_name] = convert_time(field_value,
-                                                  unit=interval_unit,
-                                                  steps_per_epoch=steps_per_epoch,
-                                                  max_training_duration=max_training_duration,
-                                                  samples_per_epoch=samples_per_epoch,
-                                                  dataset_num_tokens=dataset_num_tokens).value
 
+def _convert_time(time: Union[str, Time], state: State, ssr: float = 1.0) -> Time[int]:
+    if isinstance(time, str):
+        time = Time.from_timestring(time)
 
-@dataclass
-class SchedulerHparams(hp.Hparams, ABC):
+    if time.unit == TimeUnit.DURATION:
+        time = convert_time(time=time, unit=state.max_duration.unit, max_training_duration=state.max_duration)
+    elif ssr != 1.0:
+        time = Time(value=int(ssr * time.value), unit=time.unit)
+
+    if time.unit == TimeUnit.EPOCH:
+        time = convert_time(time=time, unit=TimeUnit.BATCH, steps_per_epoch=state.steps_per_epoch)
+
+    return time
+
+
+def step_scheduler(state: State, *, ssr: float = 1.0, step_size: Union[str, Time], gamma: float = 0.1) -> float:
```

Comment:
Alternate opinion: classes seem overkill, stateless functions are preferred for simplicity. I've always found functools.partial has always been difficult to grok/read.

Reply:
I like the idea of defining the `ComposerScheduler` type as a `Protocol`, rather than a direct function type. However, I agree with Hanlin that it's a lot simpler for users if the implementations of our schedulers are pure functions.

Reply:
Agreed; I think docstring examples can show that a scheduler can be implemented as a function, or as a class with `__call__`. Pyright should be happy with both styles :)